### PR TITLE
Docker support

### PR DESCRIPTION
**Pull request by na4ma4**
This probably needs some review since the scope creeped a bit.

This PR contains:
- devcontainer support for vscode.
- `Dockerfile` for building.
- GitHub workflows for build and push.
- changes to configuration and `view` command to handle command line parameters.
- basic `.pylintrc` config (will need to be updated for your specified code-style).

I've added a `DOCKER_CONTAINER` environment variable to the container, which causes the `config.submit()` to override with sane values (`bind_host: 0.0.0.0`, `open_in_webbrowser: False`), so when using the container users don't need to specify the bind host or not to attempt to open a webbrowser.

Python isn't my daily driver language, so it feels a bit hacky, I can split this up into multiple PRs (for devcontainer stuff, etc), but I'm expecting this to only be the first draft.

This PR addresses issue #81

### CODE REVIEW

**Comment by Owez** [comment]
Yeah sorry about that, a lot of major restructuring has been happening recently.

With #105 being the way forward with Yark, I'm not sure how to fit docker into the architecture. I want people to be able to self-host their own archives so multiple trusted people can access them, so maybe the new GUI will have a "remote connect" and docker will just contain a headless API server like what the GUI normally connects to (federated archive servers).

File: .github/workflows/docker.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      - name: Checkout
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+
+      - name: Get tag
+        id: repository
+        run: echo "tag=$(git describe --tags HEAD)" > $GITHUB_ENV
+
+      - name: Set up QEMU
```
Comment (Owez) [inline .github/workflows/docker.yml:26]: Why's QEMU being used here? (not sure about the specifics of python builds)

File: yark/archiver/config.py [hunk 2]
```diff
@@ -44,6 +45,9 @@ class Config:
     comments: bool = False
     format: Optional[str] = None
     proxy: Optional[str] = None
+    bind_host: Optional[str] = None
+    bind_port: int = 7667
+    open_in_webbrowser: bool = True
```
Comment (Owez) [inline yark/archiver/config.py:50]: Can you rename this to `headless` and have it be False by default?

File: yark/archiver/config.py [hunk 3]
```diff
@@ -70,6 +74,11 @@ def submit(self) -> None:
             )
             self.skip_download = True
 
+        # If running under docker, override some logical values
+        if os.environ.get('DOCKER_CONTAINER') == '1':
```
Comment (Owez) [inline yark/archiver/config.py:78]: `"` is the general project styling right now

File: yark/archiver/config.py [hunk 4]
```diff
@@ -110,3 +119,13 @@ def settings_md(self) -> YtDlpSettings:
 
         # Return
         return settings
+
+    def browser_url(self, archive_name: Optional[str]) -> str:
```
Comment (Owez) [inline yark/archiver/config.py:123]: Small one-line docstring here to explain what this with some comments below on main blocks

File: yark/cli.py [hunk 5]
```diff
@@ -205,35 +206,95 @@ def parse_maximum_int(config_arg: str) -> int:
...
+                elif config_arg.startswith("--port="):
+                    config.bind_port = parse_port_int(config_arg)
+
+                # Disable open in webbrowser behaviour
+                elif config_arg.startswith("--disable-webbrowser"):
+                    config.open_in_webbrowser = False
+
+                # Unknown argument
+                else:
+                    # if config_idx is 1 (first parameter could be name of archive)
```
Comment (Owez) [inline yark/cli.py:264]: Can you capitalize this comment and add one to the `else` block too

File: yark/cli.py [hunk 6]
```diff
@@ -205,35 +206,95 @@ def parse_maximum_int(config_arg: str) -> int:
...
+                        print(HELP, file=sys.stderr)
+                        _log_err(
+                            f"\nError: Unknown configuration '{config_arg}' provided for archive view"
+                        )
+                        sys.exit(1)
+
+        # Submit config settings
+        config.submit()
+
+        if config.open_in_webbrowser:
```
Comment (Owez) [inline yark/cli.py:277]: Best to move this to a new function w/ docstring + flow comments

File: yark/cli.py [hunk 7]
```diff
@@ -205,35 +206,95 @@ def parse_maximum_int(config_arg: str) -> int:
...
+            if archive_name is not None:
+                try:
+                    if not Path(archive_name).exists():
+                        raise ArchiveNotFoundException("Archive doesn't exist, please make sure you typed it's name correctly!")
+
+                    url = config.browser_url(archive_name)
+
+                    print(f"Starting viewer for {url}..")
```
Comment (Owez) [inline yark/cli.py:285]: The convention inside of the project is that archives are `path`, as `archive_name` is available here it's prefered

File: .github/workflows/docker.yml (hunk 1 above)
Comment (na4ma4) [reply .github/workflows/docker.yml:26]: To build for platforms other than amd64 qemu is needed first, I've added arm64 to the list of platforms for raspberry pi to be able to run the containers.

For reference it's in the [docker documentation for buildx actions](https://github.com/docker/setup-buildx-action#usage).

I should preserve the comments from the example.

File: yark/cli.py (hunk 6 above)
Comment (na4ma4) [reply yark/cli.py:277]: I did have an issue reusing `ArchiveNotFoundException` that it would create cyclic references if I used it in the config method.

Could we move the exceptions to a separate directory ?

File: yark/cli.py (hunk 7 above)
Comment (na4ma4) [reply yark/cli.py]: I persisted `archive_name` into the config functions as it's not a full path at that point it's just the archive name, should I change those ?

File: .github/workflows/docker.yml [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        uses: docker/setup-qemu-action@v2
+
+      - name: Set up Docker Buildx
```
[the remark this answers is not in the document]
Comment (na4ma4) [reply .github/workflows/docker.yml]: I added comments, `buildx` isn't a typo, it's the [buildkit build client](https://github.com/docker/buildx) for multi-platform support.

File: .github/workflows/docker.yml (hunk 1 above)
Comment (na4ma4) [reply .github/workflows/docker.yml]: I added in comments for each step

File: yark/archiver/config.py (hunk 2 above)
Comment (na4ma4) [reply yark/archiver/config.py]: renamed to headless, also changed the cli parameter to match

File: yark/archiver/config.py (hunk 3 above)
Comment (na4ma4) [reply yark/archiver/config.py]: Updated

File: yark/cli.py (hunk 5 above)
Comment (na4ma4) [reply yark/cli.py]: Capitalised and added comment, also done for the new methods moved to config.

File: yark/cli.py (hunk 6 above)
Comment (na4ma4) [reply yark/cli.py]: I got rid of the exception, it was adding cruft, I had it there from when I wanted to throw it from the config method and catch it here.

cleaned up and moved the webbrowser opening into config.

**Comment by na4ma4** [comment]
With future being yark server/yark gui separated and federated yark servers, most of these changes are probably just short term patches until it is ready.

File: yark/cli.py (hunk 7 above)
Comment (Owez) [reply yark/cli.py]: Oh actually if you replace "for" to "at" having the url on the viewer is nice. Two ideal options:

1. `Starting viewer for [name] archive..` then `Viewer open at [url]`
2. `Starting viewer..` then `Viewer started at [url]`

Not sure if flask pops up with the url automatically

File: yark/archiver/config.py (hunk 4 above)
Comment (Owez) [reply yark/archiver/config.py:124]: Thanks :)

**Comment by Owez** [comment]
> With future being yark server/yark gui separated and federated yark servers, most of these changes are probably just short term patches until it is ready.

Yep, I'll redo everything thats needed once the project is restructured. My current thinking is:

- `yark-pages`: GUI project which connects to `yark-api` for data/management
- `yark-api`: REST API for managing and viewing archives which uses `yark` as a dependency
  - **This is what'll be in docker**
  - Also packaged inside of `yark-pages`
- `yark`: Core library
- `yark-cli`: Quick n' dirty wrapper around `yark` for basic archive management, not as important as the GUI

File: .github/workflows/docker.yml (hunk 1 above)
Comment (Owez) [reply .github/workflows/docker.yml]: Thanks! Didn't know docker could do that, good to know

File: yark/cli.py (hunk 7 above)
Comment (na4ma4) [reply yark/cli.py]: How does this look.

`headless=False`, `archive_name=foobar`
```shell
vscode ➜ /tmp $ yark view foobar
Opening viewer at http://127.0.0.1:7667/archive/foobar/videos..
Starting server at http://127.0.0.1:7667/archive/foobar/videos for foobar
 * Serving Flask app 'yark.viewer'
 * Debug mode: off
```

`headless=False`, `archive_name=`
```shell
vscode ➜ /tmp $ yark view
Opening viewer at http://127.0.0.1:7667/..
Starting server at http://127.0.0.1:7667/
 * Serving Flask app 'yark.viewer'
 * Debug mode: off
```

`headless=True`, `archive_name=foobar`
```shell
vscode ➜ /tmp $ yark view foobar --headless
Starting server at http://127.0.0.1:7667/archive/foobar/videos for foobar
 * Serving Flask app 'yark.viewer'
 * Debug mode: off
```

`headless=True`, `archive_name=`
```shell
vscode ➜ /tmp $ yark view --headless
Starting server at http://127.0.0.1:7667/
 * Serving Flask app 'yark.viewer'
 * Debug mode: off
```

**Comment by na4ma4** [comment]
It makes sense for `yark-pages` and `yark-api` to be bundled with an option to disable one or the other.

So you could start a docker container that has a volume and runs the `yark-api` but also has `yark-pages` for management.
or start a container with a volume just for `yark-api` and either ignore that `yark-pages` is bundled or disable it with config.
or start a purely ephmeral container that only has `yark-pages` for hosting the interface.

How dynamic is `yark-pages`, could it be hosted as a static site, or does it need to be told where the `yark-api` is (if not together) ?

**Comment by na4ma4** [comment]
when this PR is ready I'll rebase it before you merge it.

File: yark/cli.py (hunk 7 above)
Comment (Owez) [reply yark/cli.py]: Hmm maybe simplify it with a single line: `Starting server at [url][suffix]` with the suffix being ` for [name]` if name is present then it'll be good to go.

**Comment by Owez** [comment]
Good point, it would be really nice to download that single binary and be able to run the api then and there. I do want Docker deployments to be the main way that people deploy headless archives for remote connections going forward.

I think yark-pages is for now limited to being a desktop application. I ideally wanted yark-pages to be a full web-app and it could do that because it's basically a static SvelteKit (think React/Angular) app put inside of the fancy tauri webview. Unfortunately tauri is in an early state and it makes it so yark-pages has to use some of its native api for file stuff which is hard to seperate.

TL;DR:
- bundle -pages and -api together
- api can be ran headless from the bundle
- -pages can't support web at the moment
- docker is good

**Comment by na4ma4** [comment]
Why does `yark-pages` need to do file stuff though, wouldn't any file stuff be the job of the API ?

File: yark/cli.py (hunk 7 above)
Comment (na4ma4) [reply yark/cli.py]: It'd probably be better to just put spaces between the [url] and the ellipsis.
eg. `Starting server at [url][suffix] ...` and then remove the "Opening" for the webbrowser.

Where suffix can be empty or ` for [name]`

Examples:

```shell
vscode ➜ /tmp $ yark view foobar
Starting server at http://127.0.0.1:7667/archive/foobar/videos for foobar ...
 * Serving Flask app 'yark.viewer'
 * Debug mode: off
vscode ➜ /tmp $ yark view
Starting server at http://127.0.0.1:7667/ ...
 * Serving Flask app 'yark.viewer'
 * Debug mode: off
vscode ➜ /tmp $ yark view foobar --headless
Starting server at http://127.0.0.1:7667/archive/foobar/videos for foobar ...
 * Serving Flask app 'yark.viewer'
 * Debug mode: off
vscode ➜ /tmp $ yark view --headless
Starting server at http://127.0.0.1:7667/ ...
 * Serving Flask app 'yark.viewer'
 * Debug mode: off
```

Comment (Owez) [reply yark/cli.py]: The styling convention which I like is `x..` instead of `x ...`

Python-y pseudocode for the address one:

```python
def loading_msg(url:str, archive_name:str|None=None):
    msg = f"Starting archive at {url} address"
    if archive_name is not None:
        msg += f" for {archive_name} archive"
    return f"{msg}.."
```

**Comment by Owez** [comment]
> Why does `yark-pages` need to do file stuff though, wouldn't any file stuff be the job of the API ?

When users want to load an archive that's not on the list of recent archives they need the file browser to popup so that they can select their archive. This needs tauri fs stuff because the webview returns `C:\fake_path\x` if you try to do it nicely with normal html/js like what happens when you do the same thing online.

I'm hoping that tauri changes permissions for the native filesystem stuff in the webview because until then the new GUI is stuck being tauri-only :/

**Comment by na4ma4** [comment]
I'd see it as something you'd query a `yark-api` job, `yark-pages` would ask a `yark-api` "what archives do you have ?" and it would return a list.

So you could potentially point a `yark-pages` at multiple `yark-api` servers, or bundle them as one app and `yark-pages` still just asks `yark-api` to do the leg work.

File: yark/cli.py (hunk 7 above)
Comment (na4ma4) [reply yark/cli.py]: Updated, put the `loading_msg` into the `launch` define.

**Comment by Owez** [comment]
> I'd see it as something you'd query a `yark-api` job, `yark-pages` would ask a `yark-api` "what archives do you have ?" and it would return a list.
> 
> So you could potentially point a `yark-pages` at multiple `yark-api` servers, or bundle them as one app and `yark-pages` still just asks `yark-api` to do the leg work.

Yep exactly :)

image

I think of it in the way where you have CoD and if you want to play a game your client needs to connect to a server. Normally it connects to a local server it spins up itself from the server inside of it's own game files, but it can also connect to friend's servers.

**Review by Owez — APPROVED** [review]

**Comment by Owez** [comment]
> when this PR is ready I'll rebase it before you merge it.

I squash and merge into master fyi, not sure if a rebase is needed if I understand you:

image

**Comment by na4ma4** [comment]
I didn't get back to it in time, but it's all done now :)

I'll wait for 1.3.x to get more fleshed out before I start looking for things to help with.

**Comment by na4ma4** [comment]
I've used the [makefiles.dev](https://github.com/make-files/makefiles) makefiles before, `make docker` works pretty well with that (then `make docker-push`).
